src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -911,6 +930,58 @@ void addLower(MutableBigInteger addend, int n) {
         add(a);
     }
 
+    /**
+     * Shifts {@code this} of {@code n} ints to the left and adds {@code addend}.
+     * Assumes {@code n > 0} for speed.
+     */
+    void shiftAdd(MutableBigInteger addend, int n) {
+        // Fast cases
+        if (addend.intLen <= n) {
+            shiftAddDisjoint(addend, n);
+        } else if (intLen == 0) {
+            copyValue(addend);
+        } else {
+            leftShift(n << 5);
+            add(addend);
```

Comment:
This method could be further optimized by implementing the shift and sum together directly here.